ambry-tools/src/main/java/com.github.ambry/store/BlobIdTransformer.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.messageformat.BlobData;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.messageformat.MessageFormatErrorCodes;
+import com.github.ambry.messageformat.MessageFormatException;
+import com.github.ambry.messageformat.MessageFormatRecord;
+import com.github.ambry.messageformat.PutMessageFormatInputStream;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static com.github.ambry.messageformat.MessageFormatRecord.*;
+
+
+/**
+ * Transformer implementation that replaces BlobIds in messages with
+ * the converted value from the StoreKeyConverter
+ */
+public class BlobIdTransformer implements StoreCopier.Transformer {
+
+  private final StoreKeyConverter storeKeyConverter;
+  private final StoreKeyFactory storeKeyFactory;
+  private final Map<StoreKey, StoreKey> cacheMap;
+
+  public BlobIdTransformer(StoreKeyConverter storeKeyConverter, StoreKeyFactory storeKeyFactory,
+      Map<StoreKey, StoreKey> cacheMap) {
+    this.storeKeyConverter = Objects.requireNonNull(storeKeyConverter, "storeKeyConverter must not be null");
+    this.storeKeyFactory = Objects.requireNonNull(storeKeyFactory, "storeKeyFactory must not be null");
+    this.cacheMap = cacheMap;
+  }
+
+  /**
+   * Takes the input message and transforms it by possibly
+   * replacing the store key and account/container IDs
+   * with a new store key and account/container IDs
+   * @param message
+   * @return
+   * @throws Exception
+   */
+  @Override
+  public StoreCopier.Message transform(StoreCopier.Message message) throws Exception {
+    Objects.requireNonNull(message, "message must not be null");
+    Objects.requireNonNull(message.getMessageInfo(), "message's messageInfo must not be null");
+    Objects.requireNonNull(message.getStream(), "message's inputStream must not be null");
+    StoreKey oldStoreKey = message.getMessageInfo().getStoreKey();
+    StoreKey newStoreKey;
+    if (cacheMap == null) {
+      List<StoreKey> list = Collections.singletonList(oldStoreKey);
+      Map<StoreKey, StoreKey> storeKeyToStoreKey = storeKeyConverter.convert(list);
+      newStoreKey = storeKeyToStoreKey.get(oldStoreKey);
+    } else {
+      newStoreKey = cacheMap.get(oldStoreKey);
```

Comment:
How does the cacheMap get populated?

Reply:
If one knows all the IDs they're going to want converted ahead of time, they can use a `StoreKeyConverter` to do a batch convert of them and provide the results to the transformer at construction.  This is in cases where the `StoreKeyConverter` `convert()` method is slow or has overhead.